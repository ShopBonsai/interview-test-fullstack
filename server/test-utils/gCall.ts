import 'reflect-metadata';
import { graphql, GraphQLSchema } from 'graphql';
import { Maybe } from 'type-graphql';
import { createSchema } from '../utils/createSchema';
import jwt from 'jsonwebtoken';

interface Options {
  source: string;
  variableValues?: Maybe<{
    [key: string]: any;
  }>;
  userId?: string;
}

let schema: GraphQLSchema;

const token = (userId) => {
  // FORCING WRONG JWT TO TEST AUTH MIDDLEWARE
  if (userId === 'force_error') {
    return {
      authorization: `Bearer ${jwt.sign({ userId }, 'wrong key')}`,
    };
  }

  return {
    authorization: `Bearer ${jwt.sign({ userId }, process.env.APP_SECRET)}`,
  };
};

export const gCall = async ({ source, variableValues, userId }: Options) => {
  if (!schema) {
    schema = await createSchema();
  }
  return graphql({
    schema,
    source,
    variableValues,
    contextValue: {
      req: {
        headers: userId ? token(userId) : {},
      },
    },
  });
};
