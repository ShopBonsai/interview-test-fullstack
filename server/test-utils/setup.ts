import { seedDatabase } from '../utils/seed.data';

const initDbTests = async () => {
  await seedDatabase();
};

initDbTests().then(() => process.exit());
