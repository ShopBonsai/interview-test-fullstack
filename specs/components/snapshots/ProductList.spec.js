import React from 'react';
import renderer from 'react-test-renderer';
import { MockedProvider } from 'react-apollo/test-utils';

import ProductsList from '../../../client/src/components/Products';
import { mocks } from '../fixtures/productList';

describe('Product List component snapshot tests', () => {
    it('should render the loading state', () => {
        const tree = renderer
            .create(
                <MockedProvider mocks={mocks}>
                    <ProductsList />
                </MockedProvider>,
            )
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render the product list of returned merchants', async () => {
        const component = renderer.create(
            <MockedProvider mocks={mocks} addTypename={false}>
                <ProductsList />
            </MockedProvider>,
        );

        // this allows the render to happen on the next tick, giving a chance for data to load
        await new Promise((resolve) => setTimeout(resolve, 0));

        expect(component.toJSON()).toMatchSnapshot();
    });
});
