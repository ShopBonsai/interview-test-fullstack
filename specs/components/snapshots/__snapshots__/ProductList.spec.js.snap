// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product List component snapshot tests should render the loading state 1`] = `
<div>
  <div>
    <div
      style={
        Object {
          "display": "flex",
          "height": "100vh",
          "width": "100vw",
        }
      }
    >
      <div
        style={
          Object {
            "alignItems": "center",
            "borderRadius": "10px",
            "boxShadow": "5px 5px 15px rgb(136 10 142 / 40%)",
            "display": "flex",
            "flexDirection": "column",
            "margin": "auto",
            "padding": "2rem",
          }
        }
      >
        <h3
          style={
            Object {
              "marginBottom": "1rem",
            }
          }
        >
          Loading products...
        </h3>
        <div
          className="spinner-border text-primary"
          role="status"
          style={
            Object {
              "height": "3rem",
              "width": "3rem",
            }
          }
        >
          <span
            className="sr-only"
          >
            Loading...
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Product List component snapshot tests should render the product list of returned merchants 1`] = `
<div>
  <div
    className="container-fluid"
    style={
      Object {
        "borderRadius": "10px",
        "boxShadow": "5px 5px 10px rgb(136 10 142 / 40%)",
        "margin": "1rem",
        "width": "95%",
      }
    }
  >
    <div
      className="row"
    >
      <div
        className="col-sm-4"
        style={
          Object {
            "padding": "2rem",
          }
        }
      >
        <div
          className="row"
        >
          <img
            alt=" Merchant logo"
            className="icon media-object"
            src="www.url.com/logo.png"
          />
          <div
            className="text-title card-title"
          >
            Vandelay Industries
          </div>
        </div>
        <p
          className="card-text"
        >
          About 
          Vandelay Industries
          : 
          A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description 
        </p>
      </div>
      <div
        className="col-sm-7"
      >
        <div
          className="product-card media"
        >
          <a
            className="media-left"
            href="#"
          >
            <img
              alt="Product image cap"
              className="img-tile media-object"
              src="www.fakeurl.com/img.png"
            />
          </a>
          <div
            className="card-body"
          >
            <div
              className="text-title card-title"
            >
              product title
            </div>
            <div
              className="card-title"
            >
              Price: $
              19.00
            </div>
            <div
              className="card-subtitle"
            >
              Color: 
              #424242
            </div>
            <div
              className="card-subtitle"
            >
              Size: 
              sm
            </div>
            <div
              className="card-subtitle"
            >
              Quantity: 
              5
            </div>
            <p
              className="card-text"
            >
              Details: 
              words words words words words words words words words words 
            </p>
            <button
              aria-label={null}
              className="btn btn-primary btn-lg btn-block"
              onClick={[Function]}
            >
              Buy
            </button>
          </div>
        </div>
        <div
          className="product-card media"
        >
          <a
            className="media-left"
            href="#"
          >
            <img
              alt="Product image cap"
              className="img-tile media-object"
              src="www.fakeurl.com/img.png"
            />
          </a>
          <div
            className="card-body"
          >
            <div
              className="text-title card-title"
            >
              product title
            </div>
            <div
              className="card-title"
            >
              Price: $
              19.00
            </div>
            <div
              className="card-subtitle"
            >
              Color: 
              #424242
            </div>
            <div
              className="card-subtitle"
            >
              Size: 
              sm
            </div>
            <div
              className="card-subtitle"
            >
              Quantity: 
              5
            </div>
            <p
              className="card-text"
            >
              Details: 
              words words words words words words words words words words 
            </p>
            <button
              aria-label={null}
              className="btn btn-primary btn-lg btn-block"
              onClick={[Function]}
            >
              Buy
            </button>
          </div>
        </div>
        <div
          className="product-card media"
        >
          <a
            className="media-left"
            href="#"
          >
            <img
              alt="Product image cap"
              className="img-tile media-object"
              src="www.fakeurl.com/img.png"
            />
          </a>
          <div
            className="card-body"
          >
            <div
              className="text-title card-title"
            >
              product title
            </div>
            <div
              className="card-title"
            >
              Price: $
              19.00
            </div>
            <div
              className="card-subtitle"
            >
              Color: 
              #424242
            </div>
            <div
              className="card-subtitle"
            >
              Size: 
              sm
            </div>
            <div
              className="card-subtitle"
            >
              Quantity: 
              5
            </div>
            <p
              className="card-text"
            >
              Details: 
              words words words words words words words words words words 
            </p>
            <button
              aria-label={null}
              className="btn btn-primary btn-lg btn-block"
              onClick={[Function]}
            >
              Buy
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="container-fluid"
    style={
      Object {
        "borderRadius": "10px",
        "boxShadow": "5px 5px 10px rgb(136 10 142 / 40%)",
        "margin": "1rem",
        "width": "95%",
      }
    }
  >
    <div
      className="row"
    >
      <div
        className="col-sm-4"
        style={
          Object {
            "padding": "2rem",
          }
        }
      >
        <div
          className="row"
        >
          <img
            alt=" Merchant logo"
            className="icon media-object"
            src="www.url.com/logo.png"
          />
          <div
            className="text-title card-title"
          >
            Vandelay Industries
          </div>
        </div>
        <p
          className="card-text"
        >
          About 
          Vandelay Industries
          : 
          A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description 
        </p>
      </div>
      <div
        className="col-sm-7"
      >
        <div
          className="product-card media"
        >
          <a
            className="media-left"
            href="#"
          >
            <img
              alt="Product image cap"
              className="img-tile media-object"
              src="www.fakeurl.com/img.png"
            />
          </a>
          <div
            className="card-body"
          >
            <div
              className="text-title card-title"
            >
              product title
            </div>
            <div
              className="card-title"
            >
              Price: $
              19.00
            </div>
            <div
              className="card-subtitle"
            >
              Color: 
              #424242
            </div>
            <div
              className="card-subtitle"
            >
              Size: 
              sm
            </div>
            <div
              className="card-subtitle"
            >
              Quantity: 
              5
            </div>
            <p
              className="card-text"
            >
              Details: 
              words words words words words words words words words words 
            </p>
            <button
              aria-label={null}
              className="btn btn-primary btn-lg btn-block"
              onClick={[Function]}
            >
              Buy
            </button>
          </div>
        </div>
        <div
          className="product-card media"
        >
          <a
            className="media-left"
            href="#"
          >
            <img
              alt="Product image cap"
              className="img-tile media-object"
              src="www.fakeurl.com/img.png"
            />
          </a>
          <div
            className="card-body"
          >
            <div
              className="text-title card-title"
            >
              product title
            </div>
            <div
              className="card-title"
            >
              Price: $
              19.00
            </div>
            <div
              className="card-subtitle"
            >
              Color: 
              #424242
            </div>
            <div
              className="card-subtitle"
            >
              Size: 
              sm
            </div>
            <div
              className="card-subtitle"
            >
              Quantity: 
              5
            </div>
            <p
              className="card-text"
            >
              Details: 
              words words words words words words words words words words 
            </p>
            <button
              aria-label={null}
              className="btn btn-primary btn-lg btn-block"
              onClick={[Function]}
            >
              Buy
            </button>
          </div>
        </div>
        <div
          className="product-card media"
        >
          <a
            className="media-left"
            href="#"
          >
            <img
              alt="Product image cap"
              className="img-tile media-object"
              src="www.fakeurl.com/img.png"
            />
          </a>
          <div
            className="card-body"
          >
            <div
              className="text-title card-title"
            >
              product title
            </div>
            <div
              className="card-title"
            >
              Price: $
              19.00
            </div>
            <div
              className="card-subtitle"
            >
              Color: 
              #424242
            </div>
            <div
              className="card-subtitle"
            >
              Size: 
              sm
            </div>
            <div
              className="card-subtitle"
            >
              Quantity: 
              5
            </div>
            <p
              className="card-text"
            >
              Details: 
              words words words words words words words words words words 
            </p>
            <button
              aria-label={null}
              className="btn btn-primary btn-lg btn-block"
              onClick={[Function]}
            >
              Buy
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="container-fluid"
    style={
      Object {
        "borderRadius": "10px",
        "boxShadow": "5px 5px 10px rgb(136 10 142 / 40%)",
        "margin": "1rem",
        "width": "95%",
      }
    }
  >
    <div
      className="row"
    >
      <div
        className="col-sm-4"
        style={
          Object {
            "padding": "2rem",
          }
        }
      >
        <div
          className="row"
        >
          <img
            alt=" Merchant logo"
            className="icon media-object"
            src="www.url.com/logo.png"
          />
          <div
            className="text-title card-title"
          >
            Vandelay Industries
          </div>
        </div>
        <p
          className="card-text"
        >
          About 
          Vandelay Industries
          : 
          A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description 
        </p>
      </div>
      <div
        className="col-sm-7"
      >
        <div
          className="product-card media"
        >
          <a
            className="media-left"
            href="#"
          >
            <img
              alt="Product image cap"
              className="img-tile media-object"
              src="www.fakeurl.com/img.png"
            />
          </a>
          <div
            className="card-body"
          >
            <div
              className="text-title card-title"
            >
              product title
            </div>
            <div
              className="card-title"
            >
              Price: $
              19.00
            </div>
            <div
              className="card-subtitle"
            >
              Color: 
              #424242
            </div>
            <div
              className="card-subtitle"
            >
              Size: 
              sm
            </div>
            <div
              className="card-subtitle"
            >
              Quantity: 
              5
            </div>
            <p
              className="card-text"
            >
              Details: 
              words words words words words words words words words words 
            </p>
            <button
              aria-label={null}
              className="btn btn-primary btn-lg btn-block"
              onClick={[Function]}
            >
              Buy
            </button>
          </div>
        </div>
        <div
          className="product-card media"
        >
          <a
            className="media-left"
            href="#"
          >
            <img
              alt="Product image cap"
              className="img-tile media-object"
              src="www.fakeurl.com/img.png"
            />
          </a>
          <div
            className="card-body"
          >
            <div
              className="text-title card-title"
            >
              product title
            </div>
            <div
              className="card-title"
            >
              Price: $
              19.00
            </div>
            <div
              className="card-subtitle"
            >
              Color: 
              #424242
            </div>
            <div
              className="card-subtitle"
            >
              Size: 
              sm
            </div>
            <div
              className="card-subtitle"
            >
              Quantity: 
              5
            </div>
            <p
              className="card-text"
            >
              Details: 
              words words words words words words words words words words 
            </p>
            <button
              aria-label={null}
              className="btn btn-primary btn-lg btn-block"
              onClick={[Function]}
            >
              Buy
            </button>
          </div>
        </div>
        <div
          className="product-card media"
        >
          <a
            className="media-left"
            href="#"
          >
            <img
              alt="Product image cap"
              className="img-tile media-object"
              src="www.fakeurl.com/img.png"
            />
          </a>
          <div
            className="card-body"
          >
            <div
              className="text-title card-title"
            >
              product title
            </div>
            <div
              className="card-title"
            >
              Price: $
              19.00
            </div>
            <div
              className="card-subtitle"
            >
              Color: 
              #424242
            </div>
            <div
              className="card-subtitle"
            >
              Size: 
              sm
            </div>
            <div
              className="card-subtitle"
            >
              Quantity: 
              5
            </div>
            <p
              className="card-text"
            >
              Details: 
              words words words words words words words words words words 
            </p>
            <button
              aria-label={null}
              className="btn btn-primary btn-lg btn-block"
              onClick={[Function]}
            >
              Buy
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="container-fluid"
    style={
      Object {
        "borderRadius": "10px",
        "boxShadow": "5px 5px 10px rgb(136 10 142 / 40%)",
        "margin": "1rem",
        "width": "95%",
      }
    }
  >
    <div
      className="row"
    >
      <div
        className="col-sm-4"
        style={
          Object {
            "padding": "2rem",
          }
        }
      >
        <div
          className="row"
        >
          <img
            alt=" Merchant logo"
            className="icon media-object"
            src="www.url.com/logo.png"
          />
          <div
            className="text-title card-title"
          >
            Vandelay Industries
          </div>
        </div>
        <p
          className="card-text"
        >
          About 
          Vandelay Industries
          : 
          A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description A longer description 
        </p>
      </div>
      <div
        className="col-sm-7"
      >
        <div
          className="product-card media"
        >
          <a
            className="media-left"
            href="#"
          >
            <img
              alt="Product image cap"
              className="img-tile media-object"
              src="www.fakeurl.com/img.png"
            />
          </a>
          <div
            className="card-body"
          >
            <div
              className="text-title card-title"
            >
              product title
            </div>
            <div
              className="card-title"
            >
              Price: $
              19.00
            </div>
            <div
              className="card-subtitle"
            >
              Color: 
              #424242
            </div>
            <div
              className="card-subtitle"
            >
              Size: 
              sm
            </div>
            <div
              className="card-subtitle"
            >
              Quantity: 
              5
            </div>
            <p
              className="card-text"
            >
              Details: 
              words words words words words words words words words words 
            </p>
            <button
              aria-label={null}
              className="btn btn-primary btn-lg btn-block"
              onClick={[Function]}
            >
              Buy
            </button>
          </div>
        </div>
        <div
          className="product-card media"
        >
          <a
            className="media-left"
            href="#"
          >
            <img
              alt="Product image cap"
              className="img-tile media-object"
              src="www.fakeurl.com/img.png"
            />
          </a>
          <div
            className="card-body"
          >
            <div
              className="text-title card-title"
            >
              product title
            </div>
            <div
              className="card-title"
            >
              Price: $
              19.00
            </div>
            <div
              className="card-subtitle"
            >
              Color: 
              #424242
            </div>
            <div
              className="card-subtitle"
            >
              Size: 
              sm
            </div>
            <div
              className="card-subtitle"
            >
              Quantity: 
              5
            </div>
            <p
              className="card-text"
            >
              Details: 
              words words words words words words words words words words 
            </p>
            <button
              aria-label={null}
              className="btn btn-primary btn-lg btn-block"
              onClick={[Function]}
            >
              Buy
            </button>
          </div>
        </div>
        <div
          className="product-card media"
        >
          <a
            className="media-left"
            href="#"
          >
            <img
              alt="Product image cap"
              className="img-tile media-object"
              src="www.fakeurl.com/img.png"
            />
          </a>
          <div
            className="card-body"
          >
            <div
              className="text-title card-title"
            >
              product title
            </div>
            <div
              className="card-title"
            >
              Price: $
              19.00
            </div>
            <div
              className="card-subtitle"
            >
              Color: 
              #424242
            </div>
            <div
              className="card-subtitle"
            >
              Size: 
              sm
            </div>
            <div
              className="card-subtitle"
            >
              Quantity: 
              5
            </div>
            <p
              className="card-text"
            >
              Details: 
              words words words words words words words words words words 
            </p>
            <button
              aria-label={null}
              className="btn btn-primary btn-lg btn-block"
              onClick={[Function]}
            >
              Buy
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
